bayes/vb.py

```diff
@@ -113,84 +115,34 @@ def plot_pdf(
         plt.show()
 
 
-class CountedEval:
-    def __init__(self, f):
-        self._f = f
-        self.n = 0
-
-    def __call__(self, *args, **kwargs):
-        self.n += 1
-        return self._f(*args, **kwargs)
-
-
-class Jacobian:
-    def __init__(self, f):
-        self._f = f
-
-    def __call__(self, xp):
+class VariationalBayesModelError:
+    def __call__(self, number_vector):
         """
-        calculate the numeric jacobian via central differences.
-
-        We _could_ change the argument x directly. For safety, we make a copy
-        at the beginning and work with that.
+        Returns a dict of type 
+            {noise_key : model_error_vector}
         """
-        x = np.copy(xp)
-
-        for iParam in range(len(x)):
-            dx = x[iParam] * 1.0e-7  # approx x0 * sqrt(machine precision)
-            if dx == 0:
-                dx = 1.0e-10
-
-            x[iParam] -= dx
-            fs0 = self._f(x)
-            x[iParam] += 2 * dx
-            fs1 = self._f(x)
-            x[iParam] -= dx
-
-            if iParam == 0:
-                # allocate jac
-                jac = []
-                for f0 in fs0:
-                    jac.append(np.empty([len(f0), len(x)]))
-
-            for i, (f0, f1) in enumerate(zip(fs0, fs1)):
-                jac[i][:, iParam] = -(f1 - f0) / (2 * dx)
-
-        return jac
-
-
-class VerifiedModelError:
-    """
-    Brings the user-provided forward model in a algorithm-compatible form.
-    * transforms a simple (numpy) vector output into a list of length 1
-    * adds a CDF jacobian, if not provided
-    """
+        raise NotImplementedError()
 
-    def __init__(self, model_error):
-        self._f = CountedEval(model_error)
-        self.n_jac = 0
-
-        try:
-            self._jac = model_error.jacobian
-        except:
-            self._jac = Jacobian(self.__call__)
+    def jacobian(self, number_vector):
+        """
+        Returns a dict of type 
+            {noise_key : d_model_error_d_number_vector_matrix}
 
-    def __call__(self, parameters):
-        f = self._f(parameters)
-        if not isinstance(f, list):
-            f = [f]
-        return f
+        By default, this is a numeric Jacobian calculated by central
+        differences.
+        """
+        return d_model_error_d_vector(self, number_vector)
 
-    def jacobian(self, parameters):
-        jac = self._jac(parameters)
-        if not isinstance(jac, list):
-            jac = [jac]
-        self.n_jac += 1
-        return jac
+class VBModelErrorWrapper(VariationalBayesModelError):
+    def __init__(self, any_me):
```

Comment:
added:
> For simple cases with only a single noise group, we want the model error to just return a vector instead of {"some_dummy_noise_key":vector}. Still, to match the VariationalBayesInterface, we use this adapter.